jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java

```diff
@@ -0,0 +1,506 @@
+        this.weightsFrequencies = new HashMap<>();
+
+        GraphPath<V, E> shortestPath = DijkstraShortestPath.findPathBetween(graph, source, sink);
+        if (shortestPath != null) {
+            candidatePaths.insert(shortestPath.getWeight(), shortestPath);
+            deviations.put(shortestPath, source);
+            weightsFrequencies.put(shortestPath.getWeight(), 1);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        return !candidatePaths.isEmpty();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public GraphPath<V, E> next() {
+        if (candidatePaths.isEmpty()) {
+            throw new NoSuchElementException();
+        }
+
+        GraphPath<V, E> path = candidatePaths.deleteMin().getValue();
+        resultList.add(path);
+        double pathWeight = path.getWeight();
+        int minWeightFrequency = weightsFrequencies.get(pathWeight);
+        if (minWeightFrequency == 1) {
+            weightsFrequencies.remove(pathWeight);
+            if (candidatePaths.isEmpty()) {
+                amountOfCandidatesWithMinimumWeight = 0;
+            } else {
+                amountOfCandidatesWithMinimumWeight = weightsFrequencies.get(candidatePaths.findMin().getKey());
+            }
+        } else {
+            weightsFrequencies.put(pathWeight, minWeightFrequency - 1);
+        }
+
+        addDeviations(path);
+
+        return path;
+    }
+
+    /**
+     * Builds unique loopless deviations from the given path
+     * in the {@link #graph}. First receives the deviation
+     * vertex of the current path as well as sets of vertices
+     * and edges to be masked during the computations. Then
+     * creates an instance of the {@link MaskSubgraph}
+     * and builds a reversed shortest paths tree starting at
+     * {@link #sink} in it. Finally builds new paths by deviating
+     * from the vertices of the provided {@code path}.
+     * <p>
+     * For more information on this step refer to the article
+     * with the original description of the algorithm.
+     *
+     * @param path path to build deviations of
+     */
+    private void addDeviations(GraphPath<V, E> path) {
+        // initializations
+        V pathDeviation = deviations.get(path);
+        List<V> pathVertices = path.getVertexList();
+        int pathVerticesSize = pathVertices.size();
+        int pathDeviationIndex = pathVertices.indexOf(pathDeviation);
+
+        // receive masked vertices and edges
+        Pair<Set<V>, Set<E>> p = getMaskedVerticesAndEdges(path, pathDeviation, pathDeviationIndex);
+        Set<V> maskedVertices = p.getFirst();
+        Set<E> maskedEdges = p.getSecond();
+
+
+        // build reversed shortest paths tree
+        Graph<V, E> maskSubgraph = new MaskSubgraph<>(graph, maskedVertices::contains, maskedEdges::contains);
+        Graph<V, E> revertedMaskedGraph = new EdgeReversedGraph<>(maskSubgraph);
```

Comment:
It's better to call it reversedMaskSubgraph since it will be consistent with previous variable's name